### PR TITLE
Single deploy and stop

### PR DESCRIPTION
**Pull request by ydye**
Deploy a single service.

### CODE REVIEW

**Comment by fanyangCS** [comment]
@sterowang is added to the review. #Closed

**Comment by fanyangCS** [comment]
@mhwu is added to the review. #Closed

**Comment by fanyangCS** [comment]
>         clean_up_generated_file(service_config)

maybe you can add single service cleanup here in the next PR

---
Refers to: service-deployment/deploy.py:361 in 5f48ae4. [](commit_id = 5f48ae49c0c22fe5e589546b1189612aead81a36, deletion_comment = False)

**Comment by ydye** [comment]
@fanyangCS Sure, I will do it, in the next PR.

File: service-deployment/deploy.py [hunk 1]
```diff
@@ -223,6 +265,28 @@ def bootstrap_service(service_config):
...
+            error_msg = "failed to mkdir -p {0}".format(dst)
+            execute_shell(shell_cmd, error_msg)
+
+        shell_cmd = "cp -r {0} {1}".format(src, dst)
+        error_msg = "failed to copy {0}".format(src)
+        execute_shell(shell_cmd, error_msg)
+
+
+
 def copy_arrangement(service_config):
```
Comment (fanyangCS) [inline service-deployment/deploy.py]: >copy_arrangement [](start = 4, length = 16)

maybe to reuse copy_arrangement_service here.

File: service-deployment/deploy.py [hunk 2]
```diff
@@ -173,6 +173,21 @@ def clean_up_generated_file(service_config):
...
+        return
+
+    for template in template_list:
+        template_data = read_template("bootstrap/{0}/{1}.template".format(serv, template))
+        generate_data = generate_from_template(template_data, cluster_config)
+        write_generated_file("bootstrap/{0}/{1}".format(serv, template), generate_data)
+
+
+
 def generate_template_file(cluster_config, service_config):
```
Comment (fanyangCS) [inline service-deployment/deploy.py]: >generate_template_file [](start = 4, length = 22)

reuse generate_template_file_service?

File: service-deployment/deploy.py [hunk 3]
```diff
@@ -191,6 +206,33 @@ def generate_template_file(cluster_config, service_config):
 
 
 
+def single_service_bootstrap(serv, service_config):
```
Comment (fanyangCS) [inline service-deployment/deploy.py]: >single_service_bootstrap [](start = 4, length = 24)

what is the difference from dependency_bootstrap? do you need to consider service dependency here?

File: service-deployment/deploy.py (hunk 1 above)
Comment (ydye) [reply service-deployment/deploy.py]: Yes, I will change it.

File: service-deployment/deploy.py (hunk 2 above)
Comment (ydye) [reply service-deployment/deploy.py]: Ok.

File: service-deployment/deploy.py (hunk 3 above)
Comment (ydye) [reply service-deployment/deploy.py]: Single service startup will ignore the dependency. User should ensure their operation is in a correctly order.

Comment (ydye) [reply service-deployment/deploy.py]: dependency_bootstrap will auto-start all service in a correctly order.

Comment (abuccts) [reply service-deployment/deploy.py]: `single_service_bootstrap` can also be reused in `bootstrap_service`?

Comment (fanyangCS) [reply service-deployment/deploy.py]: got it. please add a comment here.

---
In reply to: [159383031](https://github.com/Microsoft/pai/pull/159#discussion_r159383031) [](ancestors = 159383031)

Comment (ydye) [reply service-deployment/deploy.py]: @abuccts @fanyangCS  should we run a stop script in a function?

Comment (ydye) [reply service-deployment/deploy.py]: If we reuse single startup in bootstrap_service. If user mis-use our command. Then all service will be redeploy. In current process, if service already exists in the cluster, re-executing the startscript won't have any influence. And that's the one reason why i will make the function single_service_bootstrap  ignore dependency.

Comment (fanyangCS) [reply service-deployment/deploy.py]: I understand you introduce this function for debug purpose. I prefer everything done manually in this case.

---
In reply to: [159384788](https://github.com/Microsoft/pai/pull/159#discussion_r159384788) [](ancestors = 159384788)

**Pull request merged**